### PR TITLE
[ci] remove R 3.6 macOS CI jobs

### PR DESCRIPTION
**Pull request by jameslamb**
Removes the R 3.6 macOS jobs.

See "Motivation" and https://github.com/microsoft/LightGBM/pull/5859#discussion_r1177331016 for why.

Reasons I think this is ok:

* we still have CI jobs testing the R package against R 3.6 on Windows and Linux
* R 3.6 compatibility isn't required by CRAN any more
* the last release in the 3.6 series, 3.6.3, came out 3+ years ago ([release history](https://cran.r-project.org/bin/windows/base/old/))

### Motivation

Starting recently, I've noticed that the macOS R 3.6 jobs have been failing with the following error.

> installer: Error - the package path specified was invalid: '/Users/runner/work/LightGBM/LightGBM/R.pkg'.

([example build link](https://github.com/microsoft/LightGBM/actions/runs/4803699366/jobs/8548454888))

It looks like RStudio's CRAN changed the download link for the R 3.6 mac installer to `R-3.6.3.nn.pkg`.

Screen Shot 2023-04-25 at 10 56 09 PM

https://cran.rstudio.com/bin/macosx/

### CODE REVIEW

File: .ci/test_r_package.sh
Comment (jameslamb) [inline .ci/test_r_package.sh:136]: Unfortunately after the installer URL fix, we ran into another error... CRAN seems to no longer be serving binaries for R 3.6 packages, at least on Big Sur (macOS 12.x, the version for GitHub Actions `macos-latest`).

```text
The downloaded source packages are in
    ‘/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpB5PEzc/downloaded_packages’
Warning: unable to access index for repository https://cran.rstudio.com/bin/macosx/el-capitan/contrib/3.6:
  cannot open URL 'https://cran.rstudio.com/bin/macosx/el-capitan/contrib/3.6/PACKAGES'

   packages ‘data.table’, ‘jsonlite’, ‘knitr’, ‘Matrix’, ‘R6’, ‘RhpcBLASctl’, ‘rmarkdown’, ‘testthat’ are available as source packages but not as binaries

Warning message:
packages ‘data.table’, ‘jsonlite’, ‘knitr’, ‘Matrix’, ‘R6’, ‘RhpcBLASctl’, ‘rmarkdown’, ‘testthat’ are not available (as a binary package for R version 3.6.3) 
* checking for file ‘/Users/runner/work/LightGBM/LightGBM/lightgbm_r/DESCRIPTION’ ... OK
* preparing ‘lightgbm’:
* checking DESCRIPTION meta-information ... OK
* cleaning src
Error in loadVignetteBuilder(pkgdir, TRUE) : 
  vignette builder 'knitr' not found
Execution halted
Error in .run_shell_command("R", r_build_args) : 
  Command failed with exit code: 1
Execution halted
```

The fix on this line got past that, by allowing compilation from source in those macOS R 3.6 jobs, but that then introduced new issues.

The jobs now take more than 20 minutes just to install packages, and then fail because `{Matrix}` and `{RhpcBLASctl}` seem not to be able to compile successfully with the mix of operating system, R version, and compiler.

```text
ld: warning: directory not found for option '-L/usr/local/gfortran/lib'
ld: library not found for -lgfortran
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [Matrix.so] Error 1
ERROR: compilation failed for package ‘Matrix’
...
...
...
clang -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -fopenmp -fPIC  -Wall -g -O2  -c blasctl.c -o blasctl.o
clang: error: unsupported option '-fopenmp'
make[1]: *** [blasctl.o] Error 1
ERROR: compilation failed for package ‘RhpcBLASctl’
* removing ‘/Users/runner/Rlib/RhpcBLASctl’

The downloaded source packages are in
	‘/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpKqapM2/downloaded_packages’
Warning message:
In install.packages(c("data.table", "jsonlite", "knitr", "Matrix",  :
  installation of one or more packages failed,
  probably ‘Matrix’, ‘RhpcBLASctl’
```

([buiid link](https://github.com/microsoft/LightGBM/actions/runs/4804803508/jobs/8550640794?pr=5859))

At this point, building successfully on macOS targeting R 3.6 isn't required by CRAN. See https://www.r-project.org/nosvn/R.check/r-oldrel-macos-x86_64/lightgbm-00check.html for example, where `r-oldrel` appears to be R 4.2.0.

So in some commits tomorrow, I'll propose dropping these R 3.6 on macOS jobs, and just using the `rocker/verse` docker images to test `{lightgbm}` with R 3.6 on Linux.

**Comment by jameslamb** [comment]
@shiyu1994 could you please help with this PR? In two ways:

1. approve it so we can merge it and unblock CI in the repo
2. look at the machine for the CUDA box and maybe restart it? I'm not sure what's going wrong there, but the CUDA jobs are failing with errors that seem unrelated to this PR's changes

thank you!

**Review by guolinke — APPROVED** [review]

**Comment by jameslamb** [comment]
The CUDA jobs have been stuck waiting for a matching runner for 11+ hours. I definitely think something is wrong with the machine we use to run those tests, and we need your help @shiyu1994

**Comment by jameslamb** [comment]
@shiyu1994 can you please help with this this week? Development in this repo is blocked until the CUDA jobs are fixed, and you're the only person who can access that machine used for them.

**Comment by shiyu1994** [comment]
> @shiyu1994 can you please help with this this week? Development in this repo is blocked until the CUDA jobs are fixed, and you're the only person who can access that machine used for them.

I'm checking this now.

**Comment by shiyu1994** [comment]
I've restarted the virtual machine and now can see ci jobs running in it.

**Comment by jameslamb** [comment]
🙌🏻 thank you!!!

**Pull request merged**
